src/fastpath/fastpath.c

```diff
@@ -6,15 +6,275 @@
+                /* continue executing signaller */
+                tcbSchedEnqueue(dest);
+                restore_user_context();
+                UNREACHABLE();
+            }
+#ifdef CONFIG_VTX
+        } else if (dest && thread_state_ptr_get_tsType(&dest->tcbState) == ThreadState_RunningVM) {
+            if (tcb->tcbAffinity != getCurrentCPUIndex()) {
+                ntfn_set_active(ntfnPtr, badge);
+                doRemoteVMCheckBoundNotification(tcb->tcbAffinity, tcb);
+                restore_user_context();
+                UNREACHABLE();
+            } else {
+                Arch_leaveVMAsyncTransfer(tcb);
+                if (NODE_STATE(ksCurThread)->tcbPriority >= dest->tcbPriority) {
+#ifdef CONFIG_BENCHMARK_TRACK_KERNEL_ENTRIES
+                    ksKernelEntry.is_fastpath = true;
+#endif
+                    setRegister(dest, badgeRegister, badge);
+                    thread_state_ptr_set_tsType_np(&dest->tcbState, ThreadState_Running);
+                    /* continue executing signaller */
+                    tcbSchedEnqueue(dest);
+                    restore_user_context();
+                    UNREACHABLE();
+                }
+            }
+#endif
+        } else {
+            ntfn_set_active(ntfnPtr, badge);
+            restore_user_context();
+            UNREACHABLE();
+        }
+        break;
+    }
+    case NtfnState_Waiting: {
+        maybeDonateSchedContext(dest, ntfnPtr);
+
+        if (!dest->tcbSchedContext) {
+            slowpath(SysSend);
+        }
+
+        // Anna
+        notification_ptr_set_ntfnQueue_head_np(ntfnPtr, TCB_REF(dest->tcbEPNext));
+        if (unlikely(dest->tcbEPNext)) {
+            dest->tcbEPNext->tcbEPPrev = NULL;
+        } else {
+            notification_ptr_mset_ntfnQueue_tail_state(ntfnPtr, 0, NtfnState_Idle);
+        }
+
+        if (NODE_STATE(ksCurThread)->tcbPriority >= dest->tcbPriority) {
+#ifdef CONFIG_BENCHMARK_TRACK_KERNEL_ENTRIES
+            ksKernelEntry.is_fastpath = true;
+#endif
+            setRegister(dest, badgeRegister, badge);
+            thread_state_ptr_set_tsType_np(&dest->tcbState, ThreadState_Running);
+            SCHED_ENQUEUE(dest);
+            restore_user_context();
+            UNREACHABLE();
+        }
+        break;
+    }
+    }
+
+    /* The only way to get here is if priority of dest is higher than priority of current thread.
+     In this case, we try and switch directly to the dest thread */
+
+    /* Get destination thread VTable */
+    newVTable = TCB_PTR_CTE_PTR(dest, tcbVTable)->cap;
+
+    /* Ensure that the destination has a valid VTable. */
+    if (unlikely(!isValidVTableRoot_fp(newVTable))) {
+        slowpath(SysSend);
+    }
+
+    /* Get vspace root. */
+    cap_pd = cap_vtable_cap_get_vspace_root_fp(newVTable);
+
+#ifdef CONFIG_ARCH_AARCH32
+    /* Get HW ASID */
+    stored_hw_asid = cap_pd[PD_ASID_SLOT];
+#endif
+
+#ifdef CONFIG_ARCH_X86_64
```

Comment:
why not make this a `#elseif` cascade here for all the architectures?

Reply:
Not entirely sure to be honest. The other fastpaths (call and replyrecv) were implemented in this way so I  kept the same style. Maybe to keep them all as `#ifdef` as opposed to `#elseif defined`?

Reply:
Good point. But looking at the code it seems a good idea to make clear this is architecture specific and might get wrapped is a static inline function do get rid of the #ifdef completely